griptape/drivers/image_generation/leonardo_image_generation_driver.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional
+from attr import field, define, Factory
+from griptape.artifacts import ImageArtifact
+from griptape.drivers import BaseImageGenerationDriver
+
+
+@define
+class LeonardoImageGenerationDriver(BaseImageGenerationDriver):
+    """Driver for the Leonardo image generation API.
+
+    Attributes:
+        model: The ID of the model to use when generating images.
+        api_key: The API key to use when making requests to the Leonardo API.
+        requests_session: The requests session to use when making requests to the Leonardo API.
+        api_base: The base URL of the Leonardo API.
+        max_attempts: The maximum number of times to poll the Leonardo API for a completed image.
+        image_width: The width of the generated image in the range [32, 1024] and divisible by 8.
+        image_height: The height of the generated image in the range [32, 1024] and divisible by 8.
+
+    Details on Leonardo image generation parameters can be found here:
+    https://docs.leonardo.ai/reference/creategeneration
+    """
+
+    api_key: str = field(default=Factory(lambda: os.environ.get("LEONARDO_API_KEY")), kw_only=True)
+    requests_session: requests.Session = field(default=Factory(lambda: requests.Session()), kw_only=True)
+    api_base: str = "https://cloud.leonardo.ai/api/rest/v1"
+    max_attempts: int = field(default=10, kw_only=True)
+    image_width: int = field(default=512, kw_only=True)
+    image_height: int = field(default=512, kw_only=True)
+
+    def generate_image(
+        self, prompts: list[str], negative_prompts: Optional[list[str]] = None, **kwargs
+    ) -> ImageArtifact:
+        if negative_prompts is None:
+            negative_prompts = []
+
+        prompt = ", ".join(prompts)
+        negative_prompt = ", ".join(negative_prompts)
+
+        generation_id = self._create_generation(prompt=prompt, negative_prompt=negative_prompt)
+        image_url = self._get_image_url(generation_id=generation_id)
+        image_data = self._download_image(url=image_url)
+
+        return ImageArtifact(
+            value=image_data,
+            mime_type="image/png",
+            width=self.image_width,
+            height=self.image_height,
+            model=self.model,
+            prompt=prompt,
+        )
+
+    def _create_generation(self, prompt: str, negative_prompt: str):
+        response = self.requests_session.post(
+            url=f"{self.api_base}/generations",
+            headers={"Authorization": f"Bearer {self.api_key}"},
+            json={
+                "prompt": prompt,
+                "negative_prompt": negative_prompt,
+                "width": self.image_width,
+                "height": self.image_height,
+                "num_images": 1,
+                "modelId": self.model,
+            },
+        ).json()
+
+        return response["sdGenerationJob"]["generationId"]
+
+    def _get_image_url(self, generation_id: str):
+        for attempt in range(self.max_attempts):
```

Comment:
This retry logic should probably be moved to `BaseImageGenerationDriver` like we have in `BasePromptDriver`.

Reply:
That was my initial approach, but I found the log output from that to be overwhelming. We'll naturally expect a few request retries here to poll until the image generation is complete (or has failed), depending on service load and the request parameters and we shouldn't emit warnings under the expected happy path. 

Other providers expose fully-featured SDKs that manage polling and completion events for us and just return the image data directly, so Leonardo is (at the moment) the only driver that forces us into an async model. Maybe we want to implement retries on failed generations at the base driver level that would retry the entire generation request?

Reply:
Ahh I see, I didn't realize this is just for polling. Yeah lets retry failed ones similar to how we do in `BasePromptDriver`, but we can keep this polling logic here.